// @ngx-toolkit/device/esm5/lib/device.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, Optional } from '@angular/core';
import { DeviceService } from './device.service';
import { DEVICE, USER_AGENT } from './device.token';
/**
 * @param {?=} userAgent
 * @return {?}
 */
export function deviceResolverFactory(userAgent) {
    if (!userAgent && window) {
        userAgent = window.navigator.userAgent;
    }
    return DeviceService.resolveDevice(userAgent);
}
var DeviceModule = /** @class */ (function () {
    function DeviceModule() {
    }
    /**
     * In root module to provide the DEVICE
     */
    /**
     * In root module to provide the DEVICE
     * @return {?}
     */
    DeviceModule.forRoot = /**
     * In root module to provide the DEVICE
     * @return {?}
     */
    function () {
        return {
            ngModule: DeviceModule,
            providers: [
                {
                    provide: DEVICE,
                    useFactory: deviceResolverFactory,
                    deps: [[new Optional(), USER_AGENT]]
                }
            ]
        };
    };
    DeviceModule.decorators = [
        { type: NgModule }
    ];
    return DeviceModule;
}());
export { DeviceModule };
