// @ngx-toolkit/device/esm2015/lib/device.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, Optional } from '@angular/core';
import { DeviceService } from './device.service';
import { DEVICE, USER_AGENT } from './device.token';
/**
 * @param {?=} userAgent
 * @return {?}
 */
export function deviceResolverFactory(userAgent) {
    if (!userAgent && window) {
        userAgent = window.navigator.userAgent;
    }
    return DeviceService.resolveDevice(userAgent);
}
export class DeviceModule {
    /**
     * In root module to provide the DEVICE
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: DeviceModule,
            providers: [
                {
                    provide: DEVICE,
                    useFactory: deviceResolverFactory,
                    deps: [[new Optional(), USER_AGENT]]
                }
            ]
        };
    }
}
DeviceModule.decorators = [
    { type: NgModule }
];
