// @ngx-toolkit/device/esm2015/lib/device.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const DeviceType = {
    TABLET: 0,
    MOBILE: 1,
    NORMAL: 2,
};
export { DeviceType };
DeviceType[DeviceType.TABLET] = 'TABLET';
DeviceType[DeviceType.MOBILE] = 'MOBILE';
DeviceType[DeviceType.NORMAL] = 'NORMAL';
/** @enum {number} */
const DevicePlatform = {
    ANDROID: 0,
    IOS: 1,
    UNKNOWN: 2,
};
export { DevicePlatform };
DevicePlatform[DevicePlatform.ANDROID] = 'ANDROID';
DevicePlatform[DevicePlatform.IOS] = 'IOS';
DevicePlatform[DevicePlatform.UNKNOWN] = 'UNKNOWN';
export class Device {
    /**
     * @param {?=} type
     * @param {?=} platform
     */
    constructor(type = DeviceType.NORMAL, platform = DevicePlatform.UNKNOWN) {
        this.type = type;
        this.platform = platform;
    }
    /**
     * @return {?}
     */
    isNormal() {
        return this.type === DeviceType.NORMAL;
    }
    /**
     * @return {?}
     */
    isMobile() {
        return this.type === DeviceType.MOBILE;
    }
    /**
     * @return {?}
     */
    isTablet() {
        return this.type === DeviceType.TABLET;
    }
}
if (false) {
    /** @type {?} */
    Device.prototype.type;
    /** @type {?} */
    Device.prototype.platform;
}
